PagingWithNetworkSample/app/src/main/java/com/android/example/paging/pagingwithnetwork/reddit/repository/inMemory/byPage/PageKeyedSubredditPagingSource.kt

```diff
@@ -0,0 +1,59 @@
+/*
+ * Copyright (C) 2017 The Android Open Source Project
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.android.example.paging.pagingwithnetwork.reddit.repository.inMemory.byPage
+
+import androidx.paging.LoadType.END
+import androidx.paging.LoadType.START
+import androidx.paging.PagingSource
+import androidx.paging.PagingSource.LoadResult.Page
+import com.android.example.paging.pagingwithnetwork.reddit.api.RedditApi
+import com.android.example.paging.pagingwithnetwork.reddit.vo.RedditPost
+import kotlinx.coroutines.CoroutineDispatcher
+import kotlinx.coroutines.withContext
+import java.io.IOException
+
+/**
+ * A [PagingSource] that uses the before/after keys returned in page requests.
+ *
+ * @see [com.android.example.paging.pagingwithnetwork.reddit.repository.inMemory.byItem.ItemKeyedSubredditPagingSource]
+ */
+class PageKeyedSubredditPagingSource(
+        private val redditApi: RedditApi,
+        private val subredditName: String,
+        private val networkDispatcher: CoroutineDispatcher
+) : PagingSource<String, RedditPost>() {
+    override suspend fun load(params: LoadParams<String>): LoadResult<String, RedditPost> {
+        return try {
+            val data = withContext(networkDispatcher) {
```

Comment:
retrofit does this for you by default, remove?

I realize we have a service locator, is that providing value though? now with runBlockingTest being available, what's the value?

Reply:
wow TIL

For service locator itself I think it might be worth keeping just to have a repeatable pattern to inject fake reddit api, etc.